### PR TITLE
BLOG-103 | teracy-dev - the only truly universal productive developemnt platform with Docker on macOS, Linux and Windows

### PR DESCRIPTION
**Pull request by hoatle**

### CODE REVIEW

File: source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  Vagrant default file sync mechanism on Virtualbox.
+
+- Native Docker applications offered by Docker:
+  + Slow on Mac: https://github.com/docker/for-mac/issues/77
+  + Window 10 with recent latest build only is supported
+  + Slow on Windows: https://github.com/docker/for-win/issues/188
+
+
+- File watching is broken on VirtualBox since vboxsf does not trigger the inotify file watching
+  mechanism. `nodemon`, hello?
```
Comment (hoavu) [inline source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown]: don;t understand this "`nodemon`, hello?"

File: source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+- https://github.com/codekitchen/dinghy
+- https://github.com/leighmcculloch/docker-unison
+- https://github.com/gansbrest/hodor
+- https://github.com/adlogix/docker-machine-nfs
+- others...
+
+However, those existing projects that do not really resolve all the problems or do not reflect
+our vision for a productive development platform.
+
+You could check out our [vision for teracy-dev](/2016/08/19/teracy-dev-vision/) for such the
```
Comment (hoavu) [inline source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown]: You can check out

File: source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+- https://github.com/brikis98/docker-osx-dev
+- https://github.com/blinkreaction/boot2docker-vagrant
+- https://github.com/codekitchen/dinghy
+- https://github.com/leighmcculloch/docker-unison
+- https://github.com/gansbrest/hodor
+- https://github.com/adlogix/docker-machine-nfs
+- others...
+
+However, those existing projects that do not really resolve all the problems or do not reflect
```
Comment (hoavu) [inline source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown]: remove "that"

File: source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+- "same workflow on Linux, Mac, Windows": the usage should be the same on Mac, Linux, Windows
+
+- "native speed on mounted volumes": the speed should be as fast and native as on Linux
+
+- "2-way file sync": for syncing back and forth on mounted volumes, this is mandatory for development
+
+- "file watching (inotify)": for auto reload when files change
+
+- "customization": you have full control and are encouraged to change anything that you want to
+  suite your need, no hidden stuff, no lock-in and can still upgrade seamlessly.
```
Comment (hoavu) [inline source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown]: --> suit your need

File: source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+Let's see what the same workflow means when developing a simple hello world `nodejs` application
+from scratch with teracy-dev by watching the following videos:
```
Comment (hoavu) [inline source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown]: don;t understand "from scratch"

File: source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+On Windows:
+
+<iframe width="100%" height="630" src="https://www.youtube.com/embed/2200zvxIdAs" frameborder="0" allowfullscreen></iframe>
+
+On macOS:
+
+<iframe width="100%" height="630" src="https://www.youtube.com/embed/XF-FQQKG8PA" frameborder="0" allowfullscreen></iframe>
+
+They all contain the same steps, that's what "same workflow" means.
```
Comment (hoavu) [inline source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown]: that's what "same workflow" means. --> that's the meaning of the "same workflow".

File: source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+
+`teracy-dev` is not perfect, though. There are some improvements that we'd like to address.
+
+The first thing we'd like to improve is to reduce the setup time. It takes time (around 20-30 mins)
+to download the vagrant box (Ubuntu 16.04) for the first time. This is for the first time only, however.
+After the first time, you can destroy and provision the same VM within some minutes. We could use
+some other smaller boxes (boot2docker base box, for example) but we'd like to stick to the most popular
```
Comment (hoavu) [inline source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown]: (boot2docker base box, for example) --> (for example, boot2docker base box)

File: source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown (hunk 1 above)
Comment (hoatle) [inline source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown]: developers will understand

File: source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+- No consistent universal development environment and workflow on Linux, Windows, Mac. This is a very
+  important factor for team collaboration so that anyone from any OS platforms could collaborate and help
+  each other with no differences in the development environment.
+
+- Mounted volumes on VirtualBox use vboxsf, which is extremely slow. This is true for Boot2Docker and
+  Vagrant default file sync mechanism on Virtualbox.
+
+- Native Docker applications offered by Docker:
+  + Slow on Mac: https://github.com/docker/for-mac/issues/77
+  + Window 10 with recent latest build only is supported
```
Comment (hoatle) [inline source/_posts/2016-12-20-teracy-dev-the-only-truly-universal-productive-development-platform-with-docker-on-macos-linux-and-windows.markdown]: Only recent latest verions of Window 10 ("Build 10586 or later") are supported

**Pull request merged**